hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/TwoPhaseSnapshotCommitUtility.java

```diff
@@ -0,0 +1,427 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.processor;
+
+import com.hazelcast.function.ConsumerEx;
+import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.config.ProcessingGuarantee;
+import com.hazelcast.jet.core.BroadcastKey;
+import com.hazelcast.jet.core.Inbox;
+import com.hazelcast.jet.core.Outbox;
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.core.Processor.Context;
+import com.hazelcast.jet.impl.processor.TwoPhaseSnapshotCommitUtility.TransactionId;
+import com.hazelcast.jet.impl.processor.TwoPhaseSnapshotCommitUtility.TransactionalResource;
+import com.hazelcast.jet.impl.util.LoggingUtil;
+import com.hazelcast.logging.ILogger;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Map.Entry;
+import java.util.function.Consumer;
+
+import static com.hazelcast.jet.config.ProcessingGuarantee.AT_LEAST_ONCE;
+import static com.hazelcast.jet.config.ProcessingGuarantee.EXACTLY_ONCE;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
+
+/**
+ * A base class for transaction utilities implementing different transaction
+ * strategies.
+ *
+ * @param <TXN_ID> type fo the transaction ID
+ * @param <RES> type of the transactional resource
+ */
+public abstract class TwoPhaseSnapshotCommitUtility<TXN_ID extends TransactionId,
```

Comment:
No usages anywhere. If it has some on other branches, then maybe move the code there too?